dhcpcd.8.in
.\" Copyright 2006-2008 Roy Marples
.\" All rights reserved
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Feb 20, 2008
.Dt DHCPCD 8 SMM
.Sh NAME
.Nm dhcpcd
.Nd an RFC 2131 compliant DHCP client
.Sh SYNOPSIS
.Nm
.Op Fl dknpAEGHMLNRSTY
.Op Fl c , -script Ar script
.Op Fl h , -hostname Ar hostname
.Op Fl i , -classid Ar classid
.Op Fl l , -leasetime Ar seconds
.Op Fl m , -metric Ar metric
.Op Fl r , -request Ar address
.Op Fl t , -timeout Ar seconds
.Op Fl u , -userclass Ar class
.Op Fl F , -fqdn Ar FQDN
.Op Fl I , -clientid Ar clientid
.Op Fl P , -pidfile Ar pidfile
.Ar interface
.Nm
.Fl k , -release
.Ar interface
.Nm
.Fl x , -exit
.Ar interface
.Sh DESCRIPTION
.Nm
is an implementation of the DHCP client specified in
.Rs
.%T "RFC 2131"
.Re
.Nm
gets the host information
.Po 
IP address, routes, etc
.Pc
from a DHCP server and configures the network
.Ar interface
of the
machine on which it is running. 
.Nm
will then write DNS information to
.Xr resolvconf 8 ,
if available, otherwise directly to
.Pa /etc/resolv.conf .
.Nm
will also configure
.Pa /etc/yp.conf
and
.Pa /etc/ntpd.conf
with NIS and NTP information if the DHCP server provided them.
If those file contents changed, then
.Nm
will also attempt to restart the respective services to notify them of the
change.
If the hostname is currenly blank, (null) or localhost then
.Nm
will set the hostname to the one supplied by the DHCP server, or look it up
in DNS if none supplied.
.Nm
then daemonises and waits for the lease renewal time to lapse.
Then it attempts to renew its lease and reconfigure if the new lease changes.
.Ss Local Link configuration
If
.Nm
failed to obtain a lease, it will probe for a valid IPv4LL address
.Po
aka Zeroconf, aka APIPA
.Pc .
Once obtained it will probe every 10 seconds for a DHCP server to get a
proper address.
.Pp
Even when
.Nm
obtains a proper lease, it will still add a Local Link route
.Po
165.254.0.0/16
.Pc
so that the host can communicate with clients using these addresses.
.Pp
When using IPv4LL,
.Nm
will always succeed and return a 0 exit code. To disable this behaviour, you
can use the
.Fl L , -noipv4ll
option.
.Ss Hooking into DHCP events
.Nm
will run @PREFIX@/etc/dhcpcd.sh, or the script specified by the
.Fl c , -script
option. It will set $1 to a shell compatible file that holds various
configuration settings obtained from the DHCP server and $2 to either
up, down or new depending on the state of
.Nm .
.Nm
ignores the exist code of the script.
.Ss Fine tuning
You can fine tune the behaviour of
.Nm
with the following options :-
.Bl -tag -width indent
.It Fl d , -debug
Echo debug and informational messages to the console.
Subsequent debug options stop
.Nm
from daemonising.
.It Fl h , -hostname Ar hostname
By default,
.Nm
will send the current hostname to the DHCP server so it can register in DNS.
You can use this option to specify the
.Ar hostname
sent, or an empty string to
stop any
.Ar hostname
from being sent.
.It Fl i , -classid Ar classid
Override the DHCP vendor
.Ar classid
field we send. The default is
dhcpcd-<version>.
.It Fl k , -release
This causes an existing
.Nm
process running on the
.Ar interface
to release it's lease, deconfigure the
.Ar interface
and then exit.
.It Fl l , -leasetime Ar seconds
Request a specific lease time in
.Ar seconds .
By default
.Nm
does not request any lease time and leaves the it in the hands of the
DHCP server.
.It Fl m , -metric Ar metric
Added routes will use the
.Ar metric
on systems where this is supported
.Po
presently only Linux
.Pc .
Route metrics allow the addition of routes to the same destination across
different interfaces, the lower the metric the more it is preferred.
.It Fl n , -renew
Notifies an existing
.Nm
process running on the
.Ar interface
to renew it's lease. If
.Nm
is not running, then it starts up as normal.
.It Fl p , -persistent
.Nm
normally deconfigures the
.Ar interface
and configuration when it exits.
Sometimes, this isn't desirable if for example you have root mounted over NFS.
You can use this option to stop this from happening.
.It Fl r , -request Op Ar address
.Nm
normally sends a DHCP Broadcast to find servers to offer an address.
.Nm
will then request the address used. You can use this option to skip the
broadcast step and just request an
.Ar address .
The downside is if you request
an
.Ar address
the DHCP server does not know about or the DHCP server is not
authorative, it will remain silent. In this situation, we go back to the init
state and broadcast again.
If no
.Ar address
is given then we use the first address currently assigned to the
.Ar interface .
.It Fl s , -inform Op Ar address Op / Ar cidr
Behaves exactly like
.Fl r , -request
as above, but sends a DHCP inform instead of a request. This requires the
interface to be configured first. This does not get a lease as such, just
notifies the DHCP server of the
.Ar address
we are using.
.It Fl t , -timeout Ar seconds
Timeout after
.Ar seconds ,
instead of the default 20.
A setting of 0
.Ar seconds
causes
.Nm
to wait forever to get a lease.
.It Fl u , -userclass Ar class
Tags the DHCP message with the userclass
.Ar class .
DHCP servers use this give memebers of the class DHCP options other than the
default, without having to know things like hardware address or hostname.
.If Fl F , -fqdn Ar fqdn
Requests that the DHCP server updates DNS using FQDN instead of just a
hostname. Valid values for
.Ar fqdn
are none, ptr and both.
.Nm
dhcpcd itself never does any DNS updates.
.It Fl H , --sethostname
Forces
.Nm
to set the hostname as supplied by the DHCP server. Because some OS's and users
prefer to have just the hostname, or the full FQDN more
.Fl H , --sethostname
options change the behaviour. Below is the list of possible combinations:-
.Bl -tag -width indent
.It Fl H
set the hostname to the full FQDN.
.It Fl HH
strip the domain if it matches the dns domain.
.It Fl HHH
strip the domain regardless.
.It Fl HHHH
same as
.Fl H
but force hostname lookup via DNS.
.It Fl HHHHH
same as above, but strip the domain if it matches the dns domain.
.It Fl HHHHHH
same as above, but strip the domain regardless.
.El
.It Fl I , -clientid Ar clientid
Send
.Ar clientid
as a client identifier string. If
.Ar clientid
matches a hardware address format, such as 01:00:01:02:03:04:05 then we encode
it as that, otherwise as a string. You need to specify the hardware type in
the first byte. Ethernet is 01, and the hardware address in the example is
00:01:02:03:04:05. If the
.Ar clientid
is a blank string, then we disable DUID support and use a
.Ar clientid
as shown above.
.It Fl S, -mscsr
Microsoft have their own code for Classless Static Routes
.Po
RFC 3442
.Pc .
You can use this option to request this as well as the normal CSR. Another
instace of this option only requests the Microsoft CSR to prevent DHCP message
over-running its maximum size. DHCP server administrators should update their
CSR code from the Microsoft specific one to the RFC compliant one as the
content is fully compatible.
.It Fl P, -pidfile Ar filename
Write the PID to
.Ar filename
instead of
.Nm /var/run/dhcpcd- Ns Ar interface Ns Nm .pid.
.El
.Ss Restriciting behaviour
.Nm
will try to do as much as it can by default. However, there are sometimes
situations where you don't want the things to be configured exactly how the
the DHCP server wants. Here are some option that deal with turning these bits
off.
.Bl -tag -width indent
.It Fl A , -noarp
Don't request or claim the address by ARP.
.It Fl G , -nogateway
Don't set any default routes.
.It Fl L , -noipv4ll
Don't use IPv4LL at all.

debian/changelog
dhcpcd (1:3.2.3-11+deb7u1) oldstable-security; urgency=high

  * Fix CVE-2012-6698, CVE-2012-6699, CVE-2012-6700,
    out-of-bound reads/writes and use-after-free issues with specially
    crafted DHCP messages.
    This is a forward port of the patch applied to squeeze-lts since
    wheezy uses the same upstream version. (LP: #1517226)

 -- Guido Günther <agx@sigxcpu.org>  Sun, 27 Mar 2016 15:47:43 +0200

dhcpcd (1:3.2.3-11) unstable; urgency=high

    * Security fix, remote stack overflow: CVE-2012-2152. (closes: #671265)
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 3 May 2012 14:03:12 +0000

dhcpcd (1:3.2.3-10) unstable; urgency=low

    * Fix inhibit-duid/duid-inhibit confusion in README.
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 26 Jan 2012 21:04:22 +0000

dhcpcd (1:3.2.3-9) unstable; urgency=low

    * Fix double-free crash. (closes: #632164)
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 30 Jun 2011 21:04:22 +0000

dhcpcd (1:3.2.3-8) unstable; urgency=low

    * Add --pidfile argument. (closes: #532100)
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Mon, 14 Feb 2011 21:19:12 +0000

dhcpcd (1:3.2.3-7) unstable; urgency=low

    * Fix FTBFS on GNU/kFreeBSD. (closes: #594661)
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Tue, 31 Aug 2010 21:24:13 +0000

dhcpcd (1:3.2.3-6) unstable; urgency=low

    * Make /sbin/dhcpcd and /usr/share/man/man8/dhcpcd.8 managed 
    *    as alternatives, to allow co-existance with dhcpcd5.
    * Correctly handle multiple options in wrapper script. (closes: #577846)
    * Don't override CFLAGS in the build system. (closes: #571183)
	
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 19 Aug 2010 20:41:13 +0000

dhcpcd (1:3.2.3-5) unstable; urgency=low

    * Remove conflict with dhcp3-client. (closes: #563627)
    
 -- Simon Kelley <simon@thekelleys.org.uk>  Mon, 4 Jan 2010 16:37:13 +0000

dhcpcd (1:3.2.3-4) unstable; urgency=low

    * Comment out dodgy declaration of __u64.  (closes: #544679)
    * Conflict with dhcp3-client. (closes: #485484)
    * Fix location of config file in dhcpcd man page. (closes: #485483)
    * Allow spaces in OPTIONS. (closes: #534561)
    * Remove obsolete pid-file management from wrapper 
      script, since dhcpcd-3 doesn't require it. (closes: #536864)
    
 -- Simon Kelley <simon@thekelleys.org.uk>  Tue, 17 Nov 2009 21:00:12 +0100

dhcpcd (1:3.2.3-3) unstable; urgency=low

    * Fix wrong variable used for domain in dhcpcd.sh (closes: #512296)
    
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 29 Jan 2009 20:49:18 +0100

dhcpcd (1:3.2.3-2) unstable; urgency=low

    * Patched to use correct MAC address in DHCP renewal requests.
    *	(closes: #502191)
    
 -- Simon Kelley <simon@thekelleys.org.uk>  Tue, 14 Oct 2008 20:33:38 +0100
	
dhcpcd (1:3.2.3-1.1) unstable; urgency=medium

    * Non-maintainer upload.
    * Fix bashism in preinst maintainer script (Closes: #473515)
    * Bump Standards-Version to 3.7.3.

 -- Chris Lamb <chris@chris-lamb.co.uk>  Sat, 12 Apr 2008 03:50:59 +0100

dhcpcd (1:3.2.3-1) unstable; urgency=low

    * New upstream.
    * Restore lost changelog (closes: #467083)
    * Fix FTBFS on GNU/kFreeBSD. (closes: #467005)
					
 -- Simon Kelley <simon@thekelleys.org.uk>  Wed, 27 Feb 2008 12:06:38 +0100

dhcpcd (1:3.2.2-1) unstable; urgency=low

    * New upstream.
    * Add /etc/dhcpc/duid-inhibit logic. 
					
 -- Simon Kelley <simon@thekelleys.org.uk>  Thu, 21 Feb 2008 20:14:58 +0100

dhcpcd (1:3.0.17-2) unstable; urgency=low

    * Remove SET_DOMAIN option: this has been non-functional for
      serveral years, but on 3.0.x, setting causes dhcpcd to fail
      at startup. (closes: #419868)
					
 -- Simon Kelley <simon@thekelleys.org.uk>  Wed, 18 Apr 2007 13:34:28 +0100

dhcpcd (1:3.0.17-1) unstable; urgency=low

    * New upstream.
    * Handle change of lease-change script from /etc/dhcpc/dhcpcd.exe 
      to /etc/dhcpcd.sh and the config file from /etc/dhcpc/config
      to the more standards-conforming /etc/default/dhcpcd
    * Apply patch from SVN to make .info file compatible with old scripts.
				
 -- Simon Kelley <simon@thekelleys.org.uk>  Sat, 14 Apr 2007 17:41:32 +0100

dhcpcd (1:2.0.3-1) unstable; urgency=low

    * New upstream.
				
 -- Simon Kelley <simon@thekelleys.org.uk>  Fri, 10 Mar 2006 21:43:11 +0100

dhcpcd (1:2.0.2-1) unstable; urgency=low

    * New upstream.
				

dhcpcd.c
/* 
 * dhcpcd - DHCP client daemon
 * Copyright 2006-2008 Roy Marples <roy@marples.name>
 * All rights reserved

 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

const char copyright[] = "Copyright (c) 2006-2008 Roy Marples";

#include <sys/file.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <paths.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "config.h"
#include "client.h"
#include "dhcpcd.h"
#include "dhcp.h"
#include "interface.h"
#include "logger.h"
#include "socket.h"
#include "version.h"

static int doversion = 0;
static int dohelp = 0;
#define EXTRA_OPTS
static const struct option longopts[] = {
	{"arp",         no_argument,        NULL, 'a'},
	{"script",      required_argument,  NULL, 'c'},
	{"debug",       no_argument,        NULL, 'd'},
	{"hostname",    optional_argument,  NULL, 'h'},
	{"classid",     optional_argument,  NULL, 'i'},
	{"release",     no_argument,        NULL, 'k'},
	{"leasetime",   required_argument,  NULL, 'l'},
	{"metric",      required_argument,  NULL, 'm'},
	{"renew",       no_argument,        NULL, 'n'},
	{"persistent",  no_argument,        NULL, 'p'},
	{"inform",      optional_argument,  NULL, 's'},
	{"request",     optional_argument,  NULL, 'r'},
	{"timeout",     required_argument,  NULL, 't'},
	{"userclass",   required_argument,  NULL, 'u'},
	{"exit",        no_argument,        NULL, 'x'},
	{"lastlease",   no_argument,        NULL, 'E'},
	{"fqdn",        required_argument,  NULL, 'F'},
	{"nogateway",   no_argument,        NULL, 'G'},
	{"sethostname", no_argument,        NULL, 'H'},
	{"clientid",    optional_argument,  NULL, 'I'},
	{"noipv4ll",    no_argument,        NULL, 'L'},
	{"nomtu",       no_argument,        NULL, 'M'},
	{"nontp",       no_argument,        NULL, 'N'},
	{"nodns",       no_argument,        NULL, 'R'},
	{"msscr",       no_argument,        NULL, 'S'},
	{"test",        no_argument,        NULL, 'T'},
	{"nonis",       no_argument,        NULL, 'Y'},
	{"pidfile",     required_argument,  NULL, 'P'},
	{"help",        no_argument,        &dohelp, 1},
	{"version",     no_argument,        &doversion, 1},
#ifdef THERE_IS_NO_FORK
	{"daemonised",	no_argument,        NULL, 'f'},
	{"skiproutes",  required_argument,  NULL, 'g'},
#endif
	{NULL,          0,                  NULL, 0}
};

#ifdef THERE_IS_NO_FORK
char dhcpcd[PATH_MAX];
char **dhcpcd_argv = NULL;
int dhcpcd_argc = 0;
char *dhcpcd_skiproutes = NULL;
#undef EXTRA_OPTS
#define EXTRA_OPTS "fg:"
#endif

static int atoint (const char *s)
{
	char *t;
	long n;

	errno = 0;
	n = strtol (s, &t, 0);
	if ((errno != 0 && n == 0) || s == t ||
	    (errno == ERANGE && (n == LONG_MAX || n == LONG_MIN)))
	{
		logger (LOG_ERR, "`%s' out of range", s);
		return (-1);	
	}

	return ((int) n);
}

static pid_t read_pid (const char *pidfile)
{
	FILE *fp;
	pid_t pid = 0;

	if ((fp = fopen (pidfile, "r")) == NULL) {
		errno = ENOENT;
		return 0;
	}

	fscanf (fp, "%d", &pid);
	fclose (fp);

	return (pid);
}

static void usage (void)
{
	printf ("usage: "PACKAGE" [-adknpEGHMNRSTY] [-c script] [-h hostname] [-i classID]\n"
		"              [-l leasetime] [-m metric] [-r ipaddress] [-s ipaddress]\n"
		"              [-t timeout] [-u userclass] [-F none | ptr | both]\n"
		"              [-I clientID] [-P pidfile] <interface>\n");
}

int main(int argc, char **argv)
{
	options_t *options;
	int userclasses = 0;
	int opt;
	int option_index = 0;
	char *prefix;
	pid_t pid;
	int debug = 0;
	int i;
	int pidfd = -1;
	int sig = 0;
	int retval = EXIT_FAILURE;

	/* Close any un-needed fd's */
	for (i = getdtablesize() - 1; i >= 3; --i)
		close (i);

	openlog (PACKAGE, LOG_PID, LOG_LOCAL0);

	options = xzalloc (sizeof (*options));
	options->script = (char *) DEFAULT_SCRIPT;
	snprintf (options->classid, CLASS_ID_MAX_LEN, "%s %s",
		  PACKAGE, VERSION);

	options->doarp = true;
	options->dodns = true;
	options->domtu = true;
	options->donis = true;
	options->dontp = true;
	options->dogateway = true;
	options->daemonise = true;
	options->doinform = false;
	options->doipv4ll = true;
	options->doduid = true;
	options->timeout = DEFAULT_TIMEOUT;
	snprintf(options->pidfile, sizeof (options->pidfile), "none");

	gethostname (options->hostname, sizeof (options->hostname));
	if (strcmp (options->hostname, "(none)") == 0 ||
	    strcmp (options->hostname, "localhost") == 0)
		memset (options->hostname, 0, sizeof (options->hostname));

	/* Don't set any optional arguments here so we retain POSIX
	 * compatibility with getopt */
	while ((opt = getopt_long(argc, argv, EXTRA_OPTS
				  "c:dh:i:kl:m:npr:s:t:u:xAEF:GHI:LMNRSTYP:",
				  longopts, &option_index)) != -1)
	{
		switch (opt) {
			case 0:
				if (longopts[option_index].flag)
					break;
				logger (LOG_ERR,
					"option `%s' should set a flag",
					longopts[option_index].name);
				goto abort;
			case 'c':
				options->script = optarg;
				break;
			case 'd':
				debug++;
				switch (debug) {
					case 1:
						setloglevel (LOG_DEBUG);
						break;
					case 2:
						options->daemonise = false;
						break;
				}
				break;
#ifdef THERE_IS_NO_FORK
			case 'f':
				options->daemonised = true;
				close_fds ();
				break;
			case 'g':
				dhcpcd_skiproutes = xstrdup (optarg);
				break;
#endif
			case 'h':
				if (! optarg)
					*options->hostname = '\0';
				else if (strlen (optarg) > MAXHOSTNAMELEN) {
					logger (LOG_ERR,
						"`%s' too long for HostName string, max is %d",
						optarg, MAXHOSTNAMELEN);
					goto abort;
				} else
					strlcpy (options->hostname, optarg,
						 sizeof (options->hostname));
				break;
			case 'i':
				if (! optarg) {
					*options->classid = '\0';
				} else if (strlen (optarg) > CLASS_ID_MAX_LEN) {
					logger (LOG_ERR,
						"`%s' too long for ClassID string, max is %d",
						optarg, CLASS_ID_MAX_LEN);
					goto abort;
				} else
					strlcpy (options->classid, optarg,
						 sizeof (options->classid));
				break;
			case 'k':
				sig = SIGHUP;
				break;
			case 'l':
				if (*optarg == '-') {
					logger (LOG_ERR,
						"leasetime must be a positive value");
					goto abort;
				}
				errno = 0;
				options->leasetime = (uint32_t) strtol (optarg, NULL, 0);
				if (errno == EINVAL || errno == ERANGE) {
					logger (LOG_ERR, "`%s' out of range", optarg);
					goto abort;
				}
				break;
			case 'm':
				options->metric = atoint (optarg);
				if (options->metric < 0) {
					logger (LOG_ERR,
						"metric must be a positive value");
					goto abort;
				}
				break;
			case 'n':
				sig = SIGALRM;
				break;
			case 'p':
				options->persistent = true;
				break;
			case 's':
				options->doinform = true;
				options->doarp = false;
				if (! optarg || strlen (optarg) == 0) {
					options->request_address.s_addr = 0;
					break;
				} else {
					char *slash = strchr (optarg, '/');
					if (slash) {
						int cidr;
						/* nullify the slash, so the -r option can read the
						 * address */
						*slash++ = '\0';
						if (sscanf (slash, "%d", &cidr) != 1 ||
						    inet_cidrtoaddr (cidr, &options->request_netmask) != 0) {
							logger (LOG_ERR, "`%s' is not a valid CIDR", slash);
							goto abort;
						}
					}
				}
				/* FALLTHROUGH */
			case 'r':
				if (! options->doinform)
					options->dorequest = true;
				if (strlen (optarg) > 0 &&
				    ! inet_aton (optarg, &options->request_address))
				{ 
					logger (LOG_ERR, "`%s' is not a valid IP address", optarg);
					goto abort;
				}
				break;
			case 't':
				options->timeout = atoint (optarg);
				if (options->timeout < 0) {
					logger (LOG_ERR, "timeout must be a positive value");
					goto abort;
				}
				break;
			case 'u':
				{
					int offset = 0;
					for (i = 0; i < userclasses; i++)
						offset += (int) options->userclass[offset] + 1;
					if (offset + 1 + strlen (optarg) > USERCLASS_MAX_LEN) {
						logger (LOG_ERR, "userclass overrun, max is %d",
							USERCLASS_MAX_LEN);
						goto abort;
					}
					userclasses++;
					memcpy (options->userclass + offset + 1 , optarg, strlen (optarg));
					options->userclass[offset] = strlen (optarg);
					options->userclass_len += (strlen (optarg)) + 1;
				}
				break;
			case 'x':
				sig = SIGTERM;
				break;
			case 'A':
#ifndef ENABLE_ARP
				logger (LOG_ERR,
					"arp not compiled into dhcpcd");
				goto abort;
#endif
				options->doarp = false;
				break;
			case 'E':
#ifndef ENABLE_INFO
				logger (LOG_ERR,
					"info not compiled into dhcpcd");
				goto abort;
#endif
				options->dolastlease = true;
				break;
			case 'F':
				if (strncmp (optarg, "none", strlen (optarg)) == 0)
					options->fqdn = FQDN_NONE;
				else if (strncmp (optarg, "ptr", strlen (optarg)) == 0)
					options->fqdn = FQDN_PTR;
				else if (strncmp (optarg, "both", strlen (optarg)) == 0)
					options->fqdn = FQDN_BOTH;
				else {
					logger (LOG_ERR, "invalid value `%s' for FQDN", optarg);
					goto abort;
				}
				break;
			case 'G':
				options->dogateway = false;
				break;
			case 'H':
				options->dohostname++;
				break;
			case 'I':
				if (optarg) {
					if (strlen (optarg) > CLIENT_ID_MAX_LEN) {
						logger (LOG_ERR, "`%s' is too long for ClientID, max is %d",
							optarg, CLIENT_ID_MAX_LEN);
						goto abort;
					}
					if (strlcpy (options->clientid, optarg,
						     sizeof (options->clientid)) == 0)
						/* empty string disabled duid */
						options->doduid = false;

				} else {
					memset (options->clientid, 0, sizeof (options->clientid));
					options->doduid = false;
				}
				break;
			case 'L':
				options->doipv4ll = false;
				break;
			case 'M':
				options->domtu = false;
				break;
			case 'N':
				options->dontp = false;
				break;
			case 'R':
				options->dodns = false;
				break;
			case 'S':
				options->domscsr++;
				break;
			case 'T':
#ifndef ENABLE_INFO
				logger (LOG_ERR, "info support not compiled into dhcpcd");
				goto abort;
#endif
				options->test = true;
				options->persistent = true;
				break;
			case 'Y':
				options->donis = false;
				break;
			case 'P':
				snprintf(options->pidfile, sizeof (options->pidfile), "%s", optarg);
				break;
			case '?':
				usage ();
				goto abort;
			default:
				usage ();
				goto abort;
		}
	}
	if (doversion) {
		printf (""PACKAGE" "VERSION"\n");
		printf ("Compile time options:"
#ifdef ENABLE_ARP
			" ARP"
#endif
#ifdef ENABLE_DUID
			" DUID"
#endif
#ifdef ENABLE_INFO
			" INFO"
#endif
#ifdef ENABLE_INFO_COMPAT
			" INFO_COMPAT"
#endif
#ifdef ENABLE_IPV4LL
			" IPV4LL"
#endif
#ifdef ENABLE_NIS
			" NIS"
#endif
#ifdef ENABLE_NTP
			" NTP"
#endif
#ifdef SERVICE
			" " SERVICE
#endif
#ifdef ENABLE_RESOLVCONF
			" RESOLVCONF"
#endif
#ifdef THERE_IS_NO_FORK
			" THERE_IS_NO_FORK"
#endif
			"\n");
	}

	if (dohelp)
		usage ();

#ifdef THERE_IS_NO_FORK
	dhcpcd_argv = argv;
	dhcpcd_argc = argc;
	if (! realpath (argv[0], dhcpcd)) {
		logger (LOG_ERR, "unable to resolve the path `%s': %s",
			argv[0], strerror (errno));
		goto abort;
	}
#endif

	if (optind < argc) {
		if (strlen (argv[optind]) > IF_NAMESIZE) {
			logger (LOG_ERR,
				"`%s' too long for an interface name (max=%d)",
				argv[optind], IF_NAMESIZE);
			goto abort;
		}
		strlcpy (options->interface, argv[optind],
			 sizeof (options->interface));
	} else {
		/* If only version was requested then exit now */
		if (doversion || dohelp) {
			retval = 0;
			goto abort;
		}

		logger (LOG_ERR, "no interface specified");
		goto abort;
	}

	if (strchr (options->hostname, '.')) {
		if (options->fqdn == FQDN_DISABLE)
			options->fqdn = FQDN_BOTH;
	} else
		options->fqdn = FQDN_DISABLE;

	if (options->request_address.s_addr == 0 && options->doinform) {
		if ((options->request_address.s_addr =
		     get_address (options->interface)) != 0)
			options->keep_address = true;
	}

	if (IN_LINKLOCAL (ntohl (options->request_address.s_addr))) {
		logger (LOG_ERR,
			"you are not allowed to request a link local address");
		goto abort;
	}

	if (geteuid ())
		logger (LOG_WARNING, PACKAGE " will not work correctly unless"
			" run as root");

	prefix = xmalloc (sizeof (char) * (IF_NAMESIZE + 3));
	snprintf (prefix, IF_NAMESIZE, "%s: ", options->interface);
	setlogprefix (prefix);
	if (strcmp(options->pidfile, "none") == 0)
		snprintf (options->pidfile, sizeof (options->pidfile), PIDFILE,
			  options->interface);
	free (prefix);

	chdir ("/");
	umask (022);

	if (mkdir (INFODIR, S_IRUSR | S_IWUSR |S_IXUSR | S_IRGRP | S_IXGRP
		   | S_IROTH | S_IXOTH) && errno != EEXIST)
	{
		logger (LOG_ERR,
			"mkdir(\"%s\",0): %s\n", INFODIR, strerror (errno));
		goto abort;
	}

	if (mkdir (ETCDIR, S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP
		   | S_IROTH | S_IXOTH) && errno != EEXIST)
	{
		logger (LOG_ERR,
			"mkdir(\"%s\",0): %s\n", ETCDIR, strerror (errno));
		goto abort;
	}

	if (options->test) {
		if (options->dorequest || options->doinform) {
			logger (LOG_ERR,
				"cannot test with --inform or --request");
			goto abort;
		}

		if (options->dolastlease) {
			logger (LOG_ERR, "cannot test with --lastlease");
			goto abort;
		}

		if (sig != 0) {
			logger (LOG_ERR,
				"cannot test with --release or --renew");
			goto abort;
		}
	}

	if (sig != 0) {
		int killed = -1;
		pid = read_pid (options->pidfile);
		if (pid != 0)
			logger (LOG_INFO, "sending signal %d to pid %d",
				sig, pid);

		if (! pid || (killed = kill (pid, sig)))
			logger (sig == SIGALRM ? LOG_INFO : LOG_ERR,
				""PACKAGE" not running");

		if (pid != 0 && (sig != SIGALRM || killed != 0))
			unlink (options->pidfile);

		if (killed == 0) {
			retval = EXIT_SUCCESS;
			goto abort;
		}

		if (sig != SIGALRM)
			goto abort;	
	}

	if (! options->test && ! options->daemonised) {
		if ((pid = read_pid (options->pidfile)) > 0 &&
		    kill (pid, 0) == 0)
		{
			logger (LOG_ERR, ""PACKAGE
				" already running on pid %d (%s)",
				pid, options->pidfile);
			goto abort;
		}

		pidfd = open (options->pidfile,
			      O_WRONLY | O_CREAT | O_NONBLOCK, 0664);
		if (pidfd == -1) {
			logger (LOG_ERR, "open `%s': %s",
				options->pidfile, strerror (errno));
			goto abort;
		}

		/* Lock the file so that only one instance of dhcpcd runs
		 * on an interface */
		if (flock (pidfd, LOCK_EX | LOCK_NB) == -1) {
			logger (LOG_ERR, "flock `%s': %s",
				options->pidfile, strerror (errno));
			goto abort;
		}

		/* dhcpcd.sh should not interhit this fd */
		if ((i = fcntl (pidfd, F_GETFD, 0)) == -1 ||
		    fcntl (pidfd, F_SETFD, i | FD_CLOEXEC) == -1)
			logger (LOG_ERR, "fcntl: %s", strerror (errno));

		writepid (pidfd, getpid ());
		logger (LOG_INFO, PACKAGE " " VERSION " starting");
	}

	/* Seed random */
	srandomdev ();

	/* Massage our filters per platform */
	setup_packet_filters ();

	if (dhcp_run (options, &pidfd) == 0)
		retval = EXIT_SUCCESS;

abort:
	/* If we didn't daemonise then we need to punt the pidfile now */
	if (pidfd > -1) {
		close (pidfd);
		unlink (options->pidfile);
	}

	free (options);

#ifdef THERE_IS_NO_FORK
	/* There may have been an error before the dhcp_run function
	 * clears this, so just do it here to be safe */
	free (dhcpcd_skiproutes);
#endif

	logger (LOG_INFO, "exiting");

	exit (retval);
	/* NOTREACHED */
}
